### PR TITLE
ICU-20716 Fixing some buffer overruns in genccode

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20716
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/tools/toolutil/pkg_genc.cpp [hunk 1]
```diff
@@ -302,29 +332,41 @@ writeAssemblyCode(const char *filename, const char *destdir, const char *optEntr
...
-        memset(buffer, 0, sizeof(buffer));
-        length=T_FileStream_read(in, buffer, sizeof(buffer));
+        memset(buffer.uint32s, 0, sizeof(buffer.uint32s));
+        length=T_FileStream_read(in, buffer.uint32s, sizeof(buffer.uint32s));
         if(length==0) {
             break;
         }
-        for(i=0; i<(length/sizeof(buffer[0])); i++) {
-            column = write32(out, buffer[i], column);
+        for(i=0; i<(length/sizeof(buffer.uint32s[0])); i++) {
```
Comment (markusicu) [inline icu4c/source/tools/toolutil/pkg_genc.cpp]: This seems fishy: What if the last read sees length not as a multiple of 4?
Leave a TODO?

File: icu4c/source/tools/toolutil/pkg_genc.cpp [hunk 2]
```diff
@@ -302,29 +332,41 @@ writeAssemblyCode(const char *filename, const char *destdir, const char *optEntr
...
     T_FileStream_writeLine(out, "\n");
 
-    sprintf(bufferStr, assemblyHeader[assemblyHeaderIndex].footer,
+    count = snprintf(
+        buffer.chars, sizeof(buffer.chars),
+        assemblyHeader[assemblyHeaderIndex].footer,
         entry, entry, entry, entry,
         entry, entry, entry, entry);
-    T_FileStream_writeLine(out, bufferStr);
+    if (count + 1 > sizeof(buffer.chars)) {
```
Comment (markusicu) [inline icu4c/source/tools/toolutil/pkg_genc.cpp]: count >= sizeof

File: icu4c/source/tools/toolutil/pkg_genc.cpp [hunk 3]
```diff
@@ -354,16 +402,35 @@ writeCCode(const char *filename, const char *destdir, const char *optName, const
...
+        destdir,
+        buffer,
+        sizeof(buffer),
+        entry + uprv_strlen(entry),
+        sizeof(entry) - uprv_strlen(entry),
+        ".c",
+        optFilename);
+
     if (outFilePath != NULL) {
+        if (outFilePathCapacity < uprv_strlen(buffer) + 1) {
```
Comment (markusicu) [inline icu4c/source/tools/toolutil/pkg_genc.cpp]: strlen() >= capacity

File: icu4c/source/tools/toolutil/pkg_genc.cpp [hunk 4]
```diff
@@ -403,6 +470,10 @@ writeCCode(const char *filename, const char *destdir, const char *optName, const
         "    const char *bytes; \n"
         "} %s={ 0.0, \n",
         entry);
+    if (count + 1 > sizeof(buffer)) {
```
Comment (markusicu) [inline icu4c/source/tools/toolutil/pkg_genc.cpp]: count >= sizeof

File: icu4c/source/tools/toolutil/pkg_genc.cpp [hunk 5]
```diff
@@ -430,6 +501,10 @@ writeCCode(const char *filename, const char *destdir, const char *optName, const
         "    uint8_t bytes[%ld]; \n"
         "} %s={ 0.0, {\n",
         (long)T_FileStream_size(in), entry);
+    if (count + 1 > sizeof(buffer)) {
```
Comment (markusicu) [inline icu4c/source/tools/toolutil/pkg_genc.cpp]: count >= sizeof

File: icu4c/source/tools/toolutil/pkg_genc.cpp [hunk 6]
```diff
@@ -1061,8 +1161,21 @@ writeObjectCode(const char *filename, const char *destdir, const char *optEntryP
...
+        destdir,
+        buffer,
+        sizeof(buffer),
+        entry + entryOffset,
+        sizeof(entry) - entryOffset,
+        newSuffix,
+        optFilename);
+
     if (outFilePath != NULL) {
+        if (outFilePathCapacity < uprv_strlen(buffer) + 1) {
```
Comment (markusicu) [inline icu4c/source/tools/toolutil/pkg_genc.cpp]: strlen() >= capacity

**Comment by sffc** [comment]
PTAL

**Pull request merged**